### PR TITLE
Bug fix/internal issue #777

### PR DESCRIPTION
**Pull request by Dronplane**
### Scope & Purpose

PR fixes memory access during View optimization. Replacement attributes pointers were stored even if full substitution was not successful. But if substitution was failed  - source vector was deleted and stored pointers became invalidated. Now pointers stored only upon successful substitution.

- [x] :hankey: Bugfix (requires CHANGELOG entry)

#### Backports:
- [x] Backports required for: 3.7

#### Related Information

Also fixes https://github.com/arangodb/arangodb/issues/12995

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**

Link to Jenkins PR run:

http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12747/

### CODE REVIEW

**Review by gnusi — CHANGES_REQUESTED** [review]
let's add JS test as well

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 1]
```diff
@@ -299,7 +300,7 @@ bool attributesMatch(IResearchViewSort const& primarySort, IResearchViewStoredVa
     for (auto const& field : primarySort.fields()) {
       std::vector<std::string> postfix;
       if (latematerialized::isPrefix(field, nodeAttr.attr, false, postfix)) {
-        usedColumnsCounter[IResearchViewNode::SortColumnNumber].emplace_back(ColumnVariant(&nodeAttr.afData, fieldNum, &field, std::move(postfix)));
+        tmpUsedColumnsCounter[IResearchViewNode::SortColumnNumber].emplace_back(ColumnVariant(&nodeAttr.afData, fieldNum, &field, std::move(postfix)));
```
Comment (gnusi) [inline arangod/Aql/IResearchViewOptimizerRules.cpp:303]:
```suggestion
       tmpUsedColumnsCounter[IResearchViewNode::SortColumnNumber].emplace_back(&nodeAttr.afData, fieldNum, &field, std::move(postfix));
```

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 2]
```diff
@@ -312,7 +313,7 @@ bool attributesMatch(IResearchViewSort const& primarySort, IResearchViewStoredVa
       for (auto const& field : column.fields) {
         std::vector<std::string> postfix;
         if (latematerialized::isPrefix(field.second, nodeAttr.attr, false, postfix)) {
-          usedColumnsCounter[columnNum].emplace_back(ColumnVariant(&nodeAttr.afData, fieldNum, &field.second, std::move(postfix)));
+          tmpUsedColumnsCounter[columnNum].emplace_back(ColumnVariant(&nodeAttr.afData, fieldNum, &field.second, std::move(postfix)));
```
Comment (gnusi) [inline arangod/Aql/IResearchViewOptimizerRules.cpp:316]:
```suggestion
          tmpUsedColumnsCounter[columnNum].emplace_back(&nodeAttr.afData, fieldNum, &field.second, std::move(postfix));
```

**Review by gnusi — CHANGES_REQUESTED** [review]

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 3]
```diff
@@ -291,6 +291,7 @@ bool attributesMatch(IResearchViewSort const& primarySort, IResearchViewStoredVa
                      latematerialized::NodeWithAttrsColumn& node,
                      std::unordered_map<ptrdiff_t, std::vector<ColumnVariant>>& usedColumnsCounter) {
```
Comment (gnusi) [inline arangod/Aql/IResearchViewOptimizerRules.cpp:292]: tbh, I don't really understand why do we need `std::unoredered_map<...>` here at all given all keys are generated sequentially, ofc there might be gaps, but `std::vector<...>` IMHO should be more efficient. 
Indeed that's more a question to @iurii-i-popov, but I suggest to consider switching to `std::vector<...>`

**Review by gnusi — CHANGES_REQUESTED** [review]

File: arangod/Aql/IResearchViewOptimizerRules.cpp (hunk 3 above)
Comment (iurii-i-popov) [reply arangod/Aql/IResearchViewOptimizerRules.cpp:292]: I used to use a vector in another place. There is a problem with IResearchViewNode::SortColumnNumber. I put it to the end of the vector and got more complex code. That place was refactored and I was happy to remove the complexity there. You can also save sort column data in a separate variable.

You're welcome to change it.

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12763/

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12764/

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12766/

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12781/

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12782/

**Review by gnusi — APPROVED** [review]

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12784/

**Comment by Dronplane** [comment]
Tests blue

**Pull request merged**